### PR TITLE
KSQL-1793: First draft of Implement a User-defined Function (UDF)

### PR DESCRIPTION
**Pull request by JimGalasyn**

### CODE REVIEW

File: docs/developer-guide/implement-a-udf.rst [hunk 1]
```diff
@@ -0,0 +1,349 @@
+.. _implement-a-udf:
+
+Implement a User-defined Function (UDF)
+#######################################
+
+Prerequisites
+     - `Apache Maven <https://maven.apache.org/download.cgi>`__
+     - |cp| installed locally
```
Comment (joel-hamill) [inline docs/developer-guide/implement-a-udf.rst]: make this CP download a hyperlink to https://docs.confluent.io/current/installation/installing_cp/index.html

File: docs/developer-guide/implement-a-udf.rst [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+    path begins with ``~``.
+
+
+Create the Source and Project Files
+***********************************
+
+The following steps shows how to implement your UDF in a Java class and build
+it by defining a Maven POM file.
+
+#. Create a root directory for your UDF's source code and project files.
```
Comment (joel-hamill) [inline docs/developer-guide/implement-a-udf.rst]: i was expecting theses steps to mimic the headings that follow. shouldn't they? and if so, consider making them hyperlinks to the actual headings.

File: docs/developer-guide/implement-a-udf.rst [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
+Create the Java Source Code File
+================================
+
+The following Java code defines four overloads for a ``multiply`` function.
+The ``UdfDescription`` and ``Udf`` annotations tell KSQL Server to load the
+``Multiply`` class and look for methods to add to its list of available
+functions. For more information, see :ref:`ksql-udfs`.
+
+Copy the following code into a new file, named "Multiply.java":
```
Comment (joel-hamill) [inline docs/developer-guide/implement-a-udf.rst]: make this command syntax

```
``Multiply.java``
```

Comment (joel-hamill) [inline docs/developer-guide/implement-a-udf.rst]: should these be numbered steps?

File: docs/developer-guide/implement-a-udf.rst [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+   the package name.
+#. Create the Java source code file in the source code directory. 
+#. Create a Project Object Model (POM) file that defines how Maven builds the
+   source code.
+
+Create the Source Code Directory
+================================
+
+From the root directory for your UDF, create the source code directory. In this
+example, the package name is ``io.confluent.ksql.udfdemo``.
```
Comment (big-andy-coates) [inline docs/developer-guide/implement-a-udf.rst]: I would suggest using something that doesn't start with `io.confluent`, as people tend to cut and paste. Really, they should be using their own namespace, not ours.   Maybe use `my.company.ksql.udf.demo`?

File: docs/developer-guide/implement-a-udf.rst [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+     - `Apache Maven <https://maven.apache.org/download.cgi>`__
+     - |cp| installed locally
+     - Internet connectivity for downloading Confluent POM files
+
+Create the KSQL Extensions Directory
+************************************
+
+When you create a custom user-defined function (UDF), you implement it in Java
+and deploy it as a JAR to the KSQL extensions directory. By default, this 
+directory doesn't exist, so you need to create it and assign it in the KSQL
```
Comment (big-andy-coates) [inline docs/developer-guide/implement-a-udf.rst]: Can you raise a ticket against the tools team to have the `ext` directory created and have it set in the ksql-server.properties by default, thereby simplifying these instructions...?

File: docs/developer-guide/implement-a-udf.rst [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+``<path-to-confluent>/etc/ksql`` to add the fully qualified path to the
+``ext`` directory:
+
+.. codewithvars:: text
+
+    ksql.extension.dir=/home/my-home-dir/confluent-|release|/etc/ksql/ext
+
+.. note::
+
+    Use the fully qualified path. KSQL Server won't load extensions if the
```
Comment (big-andy-coates) [inline docs/developer-guide/implement-a-udf.rst]: It _should_ work with a path relative to the server installation directory, e.g. something like the following might work:

`../ext`


If it doesn't, we should make it work.

File: docs/developer-guide/implement-a-udf.rst [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+                                <goal>single</goal>
+                            </goals>
+                        </execution>
+                    </executions>
+                </plugin>
+            </plugins>
+        </build>
+    </project>
+
+Build the UDF Package
```
Comment (big-andy-coates) [inline docs/developer-guide/implement-a-udf.rst]: Is it worth having a step about testing the code, or at least a comment saying we _highly_ recommend the write comprehensive tests to cover their UDFs.

File: docs/developer-guide/implement-a-udf.rst [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+Use Your Custom UDF in a KSQL Query
+***********************************
+
+When your custom UDF is deployed in the KSQL extensions directory, it's loaded
+automatically when you start KSQL Server, and you can use it like you use the
+other KSQL functions.
+
+.. note::
+
+    When you make changes to your UDF code and re-deloy the JAR, you must
```
Comment (big-andy-coates) [inline docs/developer-guide/implement-a-udf.rst]: Worth mentioning `KSQL will only load UDFs on startup... so when...`

File: docs/developer-guide/implement-a-udf.rst (hunk 5 above)
Comment (big-andy-coates) [reply docs/developer-guide/implement-a-udf.rst]: Apparently @hjafarpour knows about the packaging stuff.

File: docs/developer-guide/implement-a-udf.rst (hunk 6 above)
Comment (JimGalasyn) [reply docs/developer-guide/implement-a-udf.rst]: Relative path does work! Path with `~` doesn't seem to work.

**Pull request merged**